Update the onnx Gemm op to FC/FCTransposed logic in caffe2 onnx backend

JerryShih opened the pull request:
The broadcast is used by default when the opset version is greater then 6.
#9874

JerryShih commented:
@houseroad @yinghai
How about this update?

JerryShih commented:
@houseroad 
There are some shape mismatch bugs in my patch. Please skip this reviewing request.
I will update for these bugs soon.

JerryShih commented:
@houseroad 
Could you please review these patch?

houseroad commented on caffe2/python/onnx/tests/c2_ref_test.py:
```diff
@@ -186,28 +186,122 @@ def test_gemm(self):
+        np.testing.assert_almost_equal(
+            output["Y"],
+            alpha * np.dot(A, B) + beta * C)
+        # broadcast for opset3 and 6
```
nit: an empty line

houseroad commented on caffe2/python/onnx/tests/c2_ref_test.py:
```diff
@@ -186,28 +186,122 @@ def test_gemm(self):
+            ["Y"],
+            alpha=alpha,
+            beta=beta,
+            transB=True)
```
transB=1

houseroad commented on caffe2/python/onnx/tests/c2_ref_test.py:
```diff
@@ -186,28 +186,122 @@ def test_gemm(self):
+            alpha=alpha,
+            beta=beta,
+            broadcast=1,
+            transB=True)
```
transB=1

houseroad commented on caffe2/python/onnx/tests/c2_ref_test.py:
```diff
@@ -186,28 +186,122 @@ def test_gemm(self):
+            ["Y"],
+            alpha=2.,
+            beta=3.,
+            broadcast=True)
```
broadcast=1

houseroad reviewed (approved):
Overall looks pretty good.

Nit: use int instead of bool when calling make_node, since `transA`, `transB`, and `broadcast` are all `int`.

JerryShih committed:
Update the onnx Gemm op to FC/FCTransposed logic in caffe2 onnx backend.

1. support broadcast by default if the opset version is greater than 6.
2. use FC/FCTransposed if we have broadcast and the input_c's dim
matching the FC/FCTransposed's requirement.

JerryShih committed:
Remove the broadcast test case in Gemm test.

If we have broadcast and the input_c's dim is not 1, we still can't use
FC for this case. So, remove this wrong test.

JerryShih committed:
Update Gemm test case.

1. The opset7 uses broadcast by default. Set the explicit opset version to test the different behavior.
2. Add the new FCTransposed op test case.

JerryShih committed:
Update the known_opset_version as the python onnx backend

Update the _known_opset_version value as the opset_version in 'caffe2/python/onnx/backend.py'.

JerryShih commented:
@houseroad 
Thank you. All boolean values are converted to int.
Wait for the CI test again.

JerryShih commented:
@houseroad 
CI checks have passed

bddppq commented on caffe2/onnx/backend.cc:
```diff
@@ -674,25 +674,40 @@ Caffe2Ops Caffe2Backend::CreateGemm(
+      int input_b_last_dim_index = (trans_b) ? 0 : 1;
+      if (input_c_shape.dim_size()==1 && input_c_shape.dim(0).has_dim_value() &&
+          input_b_shape.dim_size()==2 && input_b_shape.dim(input_b_last_dim_index).has_dim_value()) {
+        if (input_c_shape.dim(0).dim_value()==input_b_shape.dim(input_b_last_dim_index).dim_value()) {
```
If this check doesn't pass, it means the onnx model is *invalid*, so it's not just a matter of whether using fc or not, instead here should raise error.

houseroad replied:
It's possible... if c is one-element tensor.

bddppq replied:
@houseroad in that case, explicitly check that

houseroad replied:
Very legit. @JerryShih could you address the above comments?

JerryShih replied:
I still can't catch the idea. If there is no shape info, we can't make sure the input c fit the FC op's requirement(just like the scalar c case in previous comment). In this case, I think we should just use "MatMul+add". 
What did I miss?

@bddppq @houseroad

houseroad replied:
If only C's shape information is available and C is a 1d tensor, but B's shape is not available, we can aggressively assume that C's shape matches B's last dimension (or first dimension if transB). 1-element tensor is special, you can handle it with `MatMul + Add`. The point here is that `FC`'s performance is much better than `MatMul + Add`, so we would like to use `FC` more aggressively.

bddppq commented on caffe2/onnx/backend.cc:
```diff
@@ -674,25 +674,40 @@ Caffe2Ops Caffe2Backend::CreateGemm(
+    const auto input_b_vi_iter = ctx.value_infos().find(node.input(1));
+    if (input_c_vi_iter != ctx.value_infos().end() &&
+        input_c_vi_iter->second.type().tensor_type().has_shape() &&
+        input_b_vi_iter != ctx.value_infos().end() &&
```
This new code will only try to use FC when input_b's shape information is available, which is not desired.

houseroad replied:
I think in real world models, input b and c are usually weights, so we should be able to optimize them.

bddppq replied:
no gemm is pretty common, so you don't know whether a, b and c are inputs or weights.

houseroad replied:
@JerryShih let's be more aggressive, two things to do:
1) report error if the invalid condition appear,
2) if B's shape information is not available, always assume shapes match.

JerryShih replied:
How about these two case if c is a scalar?
a)

        A = np.random.randn(1, 3).astype(np.float32)
        B = np.random.randn(3, 1).astype(np.float32)
        C = np.random.randn(1).astype(np.float32)
        node_def = make_node(
            'Gemm',
            ['A', 'B', 'C'],
            ["Y"])
        output = c2.run_node(node_def, [A, B, C])
        np.testing.assert_almost_equal(output["Y"], np.dot(A, B) + C)

b)

        A = np.random.randn(3, 2).astype(np.float32)
        B = np.random.randn(2, 4).astype(np.float32)
        C = np.random.randn(1).astype(np.float32)
        node_def = make_node(
            'Gemm',
            ['A', 'B', 'C'],
            ["Y"])
        output = c2.run_node(node_def, [A, B, C])
        np.testing.assert_almost_equal(output["Y"], np.dot(A, B) + C)

The a) and b) could both use "MatMul+add".
The a) can use FC, but b) can't.
If there is no B's shape information, we can't recognize which case we have now.

@bddppq @houseroad

houseroad replied:
If B's shape information is not available, treating scalar as `MatMul+Add` is fine.

houseroad commented on caffe2/python/onnx/tests/c2_ref_test.py:
```diff
@@ -220,32 +314,69 @@ def test_gemm_conversion(self):
+        # will be converted to matmul + add
+        _, op_strs = backend.convert_node(node_def.SerializeToString(
+        ), [make_tensor_value_info("B", onnx.TensorProto.FLOAT, (3,2)).SerializeToString(),
+            make_tensor_value_info("C", onnx.TensorProto.FLOAT, (3,)).SerializeToString()],
```
This case may be incorrect, because C has to match A*B's dimensions. Let's set C's shape as (3,2) too.

JerryShih replied:
checked.

bddppq reviewed (dismissed):
I think what this PR is trying to improve is to use FCTransposed when trans_b is False, this is a good idea. Could you update the pull request's title & summary accordingly?

houseroad commented:
A potential follow up change (after getting this merged):


Right now, FC only accpet 1d bias, however, we can improve FC's implementation to accept 2d bias.

FC's implementation:
https://github.com/pytorch/pytorch/blob/master/caffe2/operators/fully_connected_op.h#L69

Same to FCTranspose.

When FC accepts 2d bias, we can always turn a Gemm into Scale + FC(Transpose)

JerryShih commented:
@houseroad @bddppq 
I update some comments in review page.

JerryShih commented:
@houseroad @bddppq
What do you think for my comments? Did I miss something?

JerryShih commented:
@houseroad @bddppq 
Request review again.

I update the condition for converting to FC.

1. If c is "1d" tensor and c is not a scalar, use FC.
2. If c is a scalar and b's shape is not available, skip FC.
3. If c is a scalar and a*b's last dim is 1, use FC.
4. If c is a scalar and a*b's last dim is not 1, skip FC.

In most case, we will use FC.

JerryShih commented:
@houseroad @bddppq 
Any feedback?

"pr/caffe2-py2-cuda9.0-cudnn7-windows-build — Build failed"
This test always failed, but that's not related to these patches.

houseroad commented:
I have restarted the build, don’t worry. I will review today later.

JerryShih commented:
@houseroad
It already has taken 3 days waiting for merging pr. Is it any test failed?

houseroad commented:
I am running some internal test cases. Will update it soon.

JerryShih commented:
yes, I will do that.

houseroad commented:
Awesome, if you have any question, feel free to ping on gitter.
